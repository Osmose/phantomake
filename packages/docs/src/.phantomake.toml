baseUrl = "https://mkelly.me/phantomake"
