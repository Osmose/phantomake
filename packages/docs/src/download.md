---
title: Download
---

### Binaries

You can download the latest binaries for your platform here:

- [MacOS](https://github.com/Osmose/phantomake/releases/latest/download/phantomake-macos.tar.gz)
- [Linux](https://github.com/Osmose/phantomake/releases/latest/download/phantomake-linux.tar.gz)
- [Windows](https://github.com/Osmose/phantomake/releases/latest/download/phantomake-windows.zip)

Phantomake does not currently have an update mechanismâ€”you will need to redownload the binary to update to a newer version.

Older versions of Phantomake can be downloaded from the [Releases](https://github.com/Osmose/phantomake/releases) page on Github.

### Install via Bun

If you have [Bun](https://bun.sh/) installed, you can also install Phantomake as a global package:

```sh
bun add -g https://github.com/Osmose/phantomake/releases/latest/download/source.tar.gz
```

This will add the `phantomake` command to your `PATH` and will run it on your local version of bun.

You can update to the latest version of Phantomake by re-running the command above after a new version is released.
