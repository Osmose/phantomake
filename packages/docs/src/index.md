---
title: Overview
---

<div class="overview-summary-phanto">

<div class="summary">

### What is Phantomake?

Phantomake is a static site generator with specific goals:

- Distributed as a single, standalone executable
- When run on an existing static site, will (in most cases) output the exact same site with no changes
- Can reuse common code (headers/footers/common HTML) across several pages with templates and includes
- Can do pagination (e.g. for blog listings)

</div>

<div class="phanto">
  <img src="left_phantomake.png" title="Phanto" alt="Phanto, the construction-hat-wearing ghost mascot of Phantomake." class="phanto">
</div>

</div>

<p class="hero-text">Phantomake turns files like this:</p>

<div class="side-by-side">

<div class="code-block-with-filename">

<div class="filename">ðŸ“„ /src/.templates/default.ejs</div>

```erb
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Phanto's Ghostlace Site</title>
  </head>
  <body>
    <h1>
      <%- output.file.attributes.title %>
    </h1>
    <%- output.content %>
  </body>
</html>
```

</div>

<div class="code-block-with-filename">

<div class="filename">ðŸ“„ /src/favorites.md</div>

```md
---
title: My Favorite Ghostlace Knots
---
I'd like to share with you a few of my
favorite ghostlace knots!

1. [Phanto's Secure Knot](secure.html)
2. [Reef Knot](reef.html)
3. [Berluti Knot](berluti.html)
```
</div>

</div>

<p class="hero-text">into this:<p>

<div class="code-block-with-filename">

<div class="filename">ðŸ“„ /output/favorites.html</div>

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Phanto's Ghostlace Site</title>
  </head>
  <body>
    <h1>My Favorite Ghostlace Knots</h1>
    <p>I'd like to share with you a few of my favorite ghostlace knots!</p>
    <ol>
      <li><a href="secure.html">Phanto's Secure Knot</a></li>
      <li><a href="reef.html">Reef Knot</a></li>
      <li><a href="berluti.html">Berluti Knot</a></li>
    </ol>
  </body>
</html>
```

</div>
