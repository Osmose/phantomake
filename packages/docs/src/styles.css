:root {
  --content-width: 960px;
  --content-side-padding: 16px;

  --color-black: #1a1c2c;
  --color-maroon: #5d275d;
  --color-red: #b13e53;
  --color-orange: #ef7d57;
  --color-yellow: #ffcd75;
  --color-lime: #a7f070;
  --color-green: #38b764;
  --color-teal: #257179;
  --color-navy: #29366f;
  --color-blue: #3b5dc9;
  --color-sky: #41a6f6;
  --color-cyan: #73eff7;
  --color-white: #f4f4f4;
  --color-white-grey: #e5ecf1;
  --color-light-grey: #94b0c2;
  --color-grey: #566c86;
  --color-dark-grey: #333c57;
}

* {
  box-sizing: border-box;
}

body {
  font-size: 16px;
  font-family: ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
  line-height: 1.5;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}

body .topbar {
  flex: 0 0 auto;
}

body .page {
  flex: 1;
}

body .footer {
  flex: 0 0 auto;
}

a {
  color: var(--color-blue);
}

a:visited {
  color: var(--color-navy);
}

code.hljs {
  font-size: .9em;
  margin: 2em 0;
}

code:not(.hljs) {
  font-family: ui-monospace,SFMono-Regular,SF Mono,Menlo,Consolas,Liberation Mono,monospace;
  background: var(--color-white-grey);
  padding: .2em .4em;
  font-size: 85%;
  white-space: break-spaces;
}

.topbar {
  color: var(--color-white);
  background: var(--color-dark-grey);
}

.topbar-content {
  margin: 0 auto;
  padding: 12px var(--content-side-padding);
  line-height: 48px;
  max-width: var(--content-width);
  display: flex;
  flex-direction: row;
  position: relative;
}

.topbar-title {
  flex: 1;
  margin: 0;
  font-weight: 500;
}

.topbar-title a {
  color: var(--white);
  text-decoration: none;
}

.topbar-nav {
  flex: 0 0 auto;
  font-size: .8em;
}

.topbar-nav-link {
  color: var(--color-white);
  text-transform: uppercase;
  text-decoration: none;
  font-weight: 600;
  padding: 12px;
  line-height: 48px;
}

.topbar-nav-link:hover {
  color: var(--color-lime);
}

.topbar-nav-link:visited {
  color: var(--color-white);
}

.topbar-nav-link.selected {
  color: var(--color-green);
}

.topbar-nav-link .github-icon {
  width: 32px;
  height: 32px;
  vertical-align: -10px;
}

.page-header {
  color: var(--color-white);
  background: var(--color-grey);
}

.page-title {
  margin: 0 auto;
  max-width: var(--content-width);
  padding: 12px var(--content-side-padding);
  font-weight: 300;
}

.page-content {
  margin: 12px auto;
  padding: 0 var(--content-side-padding);
  max-width: var(--content-width);
}

.footer {
  text-align: right;
  margin: 4em auto 1em;
  max-width: var(--content-width);
  padding: 0 var(--content-side-padding);
  color: var(--color-light-grey);
}

.footer a {
  color: var(--color-light-grey);
}

.page-content li {
  margin: .6em 0;
}

h3,
h4,
h5,
h6 {
  margin: 1.4em 0 .4em;
}

h3:first-of-type {
  margin-top: .5em;
}

h3 {
  /* margin: 2.5em 0 .5em; */
  margin-top: 2em;
  color: var(--color-green);
  font-size: 2.2rem;
  font-weight: 400;
}

h4 {
  color: var(--color-green);
  font-size: 1.4rem;
  font-weight: 400;
}

h5 {
  color: var(--color-green);
  font-size: 1.2rem;
  font-weight: 600;
}

h6 {
  color: var(--color-green);
  font-size: 1rem;
}

p {
  margin: 1.2em 0;
}

hr {
  border: 0;
  border-top: 1px solid var(--color-white-grey);
  margin: 2em 0;
}

hr + h3,
hr + h4,
hr + h5,
hr + h6 {
  margin-top: 1em;
}

h3 > a,
h4 > a,
h5 > a,
h6 > a {
  color: inherit;
  text-decoration: none;
}

h3 > a:visited,
h4 > a:visited,
h5 > a:visited,
h6 > a:visited {
  color: inherit;
}

figure {
  text-align: center;
  margin: 2em 0;
}

figure > img {
  background: var(--color-white-grey);
  padding: 1em;
}

figcaption {
  font-size: .9em;
  color: var(--color-grey);
}

.toc {
  padding: 1em;
  border: 1px solid var(--color-light-grey);
  background: var(--color-white-grey);
  display: inline-block;
  margin: 1em 0 0;
}

.toc-title {
  font-weight: 600;
}

.code-block-with-filename {
  margin: 2em 0;
}

.code-block-with-filename .filename {
  font-family: ui-monospace,SFMono-Regular,SF Mono,Menlo,Consolas,Liberation Mono,monospace;
  color: var(--color-white);
  background: var(--color-grey);
  padding: .6em 1em;
  font-size: 85%;
  white-space: break-spaces;
}

.code-block-with-filename pre {
  margin: 0;
}

.code-block-with-filename .hljs {
  margin: 0;
}

.hero-text {
  font-size: 2em;
  text-align: center;
  margin: 1em 0;
  font-weight: 600;
}

.side-by-side {
  display: flex;
  flex-direction: row;
  gap: 2em;
  background: var(--color-white-grey);
  padding: 2em;
}

.side-by-side > * {
  flex: 1;
  margin: 0;
  font-size: 0.9em;
}

.admonition {
  background: var(--color-white-grey);
  border: 1px solid var(--color-grey);
  padding: 1em;
  margin: 2em 0;
}

.admonition-title {
  font-weight: 600;
  color: var(--color-grey);
}

.admonition-title::before {
  content: 'ℹ️ ';
}

.admonition *:first-child {
  margin-top: 0;
}

.admonition *:last-child {
  margin-bottom: 0;
}

.overview-summary-phanto {
  display: flex;
  flex-direction: row;
}

.overview-summary-phanto .summary {
  flex: 1;
}

.overview-summary-phanto .phanto {
  flex: 0 0 auto;
  margin-top: 1em;
}

/* Show wide nav for resolutions > 800px, narrow nav for < 800px. */

.topbar-narrow-nav {
  display: none;
  transition: all .2s ease-in-out;
  height: 0;
  overflow: hidden;
}

.topbar-narrow-nav > :not(summary) {
  display: block;
}

.topbar-narrow-nav summary {
  list-style: none;
  cursor: pointer;
  position: absolute;
  top: 20px;
  right: var(--content-side-padding);
}

.topbar-narrow-nav summary:hover {
  color: var(--color-green);
}

.topbar-narrow-nav .menu-icon {
  width: 32px;
  height: 32px;
}

.topbar-narrow-nav[open] {
  height: 344px;
}

.topbar-narrow-nav[open] summary {
  color: var(--color-grey);
}

.api-container {
  display: grid;
  grid-template-columns: 200px repeat(4, 1fr);
  grid-template-rows: 1fr;
  grid-column-gap: 2em;
  grid-row-gap: 0px;
}

.api-conttainer .api-toc {
  grid-area: 1 / 1 / 2 / 2;
}

.api-toc .toc {
  display: block;
  position: sticky;
  top: 2em;
}

.api-container .api-content {
  grid-area: 1 / 2 / 2 / 3;
}

@media only screen and (prefers-reduced-motion) {
  .topbar-narrow-nav {
    transition: none;
    height: auto;
  }

  .topbar-narrow-nav[open] {
    height: auto;
  }
}

@media only screen and (max-width: 799px){
  .topbar-content {
    flex-direction: column;
  }

  .topbar-nav-link {
    padding: 8px 0;
    line-height: 32px;
    border-bottom: 1px solid var(--color-light-grey);
  }

  .topbar-nav-link:last-child {
    border-bottom: none;
  }

  .topbar-narrow-nav {
    display: block;
  }

  .topbar-wide-nav {
    display: none;
  }

  .overview-summary-phanto {
    flex-direction: column-reverse;
  }

  .overview-summary-phanto .phanto {
    text-align: center;
    margin: 0;
  }

  .side-by-side {
    flex-direction: column;
  }

  .api-container {
    display: block;
  }

  .api-toc .toc {
    position: relative;
    top: 0;
    margin: 1em 0 0;
    display: inline-block;
  }
}
