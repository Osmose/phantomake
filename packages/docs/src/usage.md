---
title: Usage
---

### How to run Phantomake

There are two ways to run Phantomake:

- If you are comfortable with using a terminal, use the [terminal command](#terminal-command).
- If you don't want to use the terminal, then try [PhantoGUI](#phantogui).

### Terminal Command

Phantomake is a terminal command. If you're unfamiliar with using a terminal, [MDN has a tutorial that can teach you the basics](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line).

Phantomake has two subcommands: `build` and `watch`. If no subcommand is given, it will default to the `build` subcommand.

#### `phantomake [build] <source_directory> <output_directory>`

The `build` subcommand runs Phantomake on a source directory, processes the source files into output, and saves that output to the output directory.

> [!NOTE]
> Phantomake will overwrite any existing files that match an output file it is writing, but it does not delete existing files in the output directory that _don't_ conflict with any output.

```sh
# Processes files in `src` and output them to the `dist` directory
phantomake ./src ./dist
```

#### `phantomake watch <source_directory>`

The `watch` subcommand makes working on a Phantomake-generated site much more convenient. It watches the source directory and automatically rebuilds on any file changes. The output is written to a temporary directory and is made viewable on a local development server. By default this is [http://localhost:8000](http://localhost:8000) but can be customized with the `--host` and `--port` options if needed.

```sh
# Watches `src` for changes and hosts the build output at http://localhost:8000
phantomake watch ./src
```

### PhantoGUI

<img src="phantogui.png" class="phantogui-screenshot" alt="Screenshot of the user interface of PhantoGUI, showing the output of a preview server.">

PhantoGUI allows you to run Phantomake without having to use the terminal. It includes:

- A tab for launching and monitoring the **Watcher** preview of your output site.
- A **Publish** tab for building and saving your output to a local directory.

PhantoGUI runs on Linux, MacOS, or Windows and can be downloaded from the [Downloads](download.html) page.
