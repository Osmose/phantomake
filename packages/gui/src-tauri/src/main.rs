// Prevents additional console window on Windows in release, DO NOT REMOVE!!
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

fn main() {
  tauri::Builder::default()
    .plugin(tauri_plugin_window_state::Builder::default().build())
    .plugin(tauri_plugin_persisted_scope::init())
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
