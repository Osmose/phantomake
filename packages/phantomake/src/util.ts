import * as nodePath from 'node:path';
import * as fs from 'node:fs/promises';

/** Recursively walk a directory tree and return the path of every individual file found. */
export async function walk(directory: string): Promise<string[]> {
  const fileNames = await fs.readdir(directory);
  const walkedDirectoryPromises = Promise.all(
    fileNames.map(async (fileName) => {
      const filePath = nodePath.join(directory, fileName);
      const stat = await fs.lstat(filePath);
      if (stat.isDirectory() && !stat.isSymbolicLink()) {
        return walk(filePath);
      }

      return filePath;
    })
  );

  return (await walkedDirectoryPromises).flat();
}
