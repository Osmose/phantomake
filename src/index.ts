import * as nodePath from 'node:path';
import * as fs from 'node:fs/promises';
import * as os from 'node:os';
import ejs from 'ejs';
import { consola } from 'consola';
import { InputFile } from './base';
import { FileContext, GlobalContext } from './context';

/** Recursively walk a directory tree and return the path of every individual file found. */
async function walk(directory: string): Promise<string[]> {
  const fileNames = await fs.readdir(directory);
  const walkedDirectoryPromises = Promise.all(
    fileNames.map(async (fileName) => {
      const filePath = nodePath.join(directory, fileName);
      const stat = await fs.lstat(filePath);
      if (stat.isDirectory() && !stat.isSymbolicLink()) {
        return walk(filePath);
      }

      return filePath;
    })
  );

  return (await walkedDirectoryPromises).flat();
}

interface Template {
  render: ejs.TemplateFunction;
  context: FileContext;
}

class Templates {
  constructor(private readonly templates: { [name: string]: Template }) {}

  apply(templateName: string | null | undefined, data?: ejs.Data | undefined) {
    const template = this.templates[templateName ?? 'default'];
    if (!template) {
      if (!templateName) {
        throw new Error(
          'Markdown files without an explicit template require a default HTML template at .templates/default.ejs'
        );
      } else {
        throw new Error(`No template found under .templates named ${templateName}.`);
      }
    }

    return template.render({ ...data, ctx: template.context });
  }

  static async fromInputFiles(globalContext: GlobalContext, inputFiles: InputFile[]) {
    const templates: { [name: string]: Template } = {};
    for (const inputFile of inputFiles) {
      if (!inputFile.isTemplate) {
        continue;
      }

      templates[inputFile.parsedRelativePath.name] = {
        render: ejs.compile(await inputFile.file.text(), {
          filename: inputFile.path,
        }),
        context: globalContext.fileContext(inputFile),
      };
    }

    return new Templates(templates);
  }
}

interface Output {
  content: string;
  file: InputFile;
  path: string;
}

async function renderOutput(inputFile: InputFile, globalContext: GlobalContext, templates: Templates): Promise<Output> {
  const output = {
    content: inputFile.body ?? '',
    file: inputFile,
    path: inputFile.relativePath,
  };
  const context = globalContext.fileContext(inputFile);

  // Apply the matching processor if one was found
  if (inputFile.processor) {
    output.path = inputFile.processor.outputPath(inputFile);
    output.content = await inputFile.processor.process(inputFile, context);
  }

  // If a template was defined in the frontmatter, apply the processed content to it
  const attributes = inputFile.attributes;
  if (attributes.template) {
    output.content = templates.apply(attributes.template, { ctx: context, output });
  }

  return output;
}

export interface PhantomakeOptions {
  logging?: boolean;

  /** URL / domain name to be prepended to absolute URLs. */
  baseUrl?: string;
}

export default async function phantomake(
  inputDirectory: string,
  outputDirectory: string,
  options: PhantomakeOptions = {}
) {
  const logger = options.logging ? consola : null;

  // Find input files and prepare for processing
  const inputPaths = await walk(inputDirectory);
  const inputFiles: InputFile[] = [];
  for (const path of inputPaths) {
    inputFiles.push(await InputFile.fromPath(inputDirectory, path));
  }
  const globalContext = new GlobalContext(inputFiles, { baseUrl: options.baseUrl });
  const templates = await Templates.fromInputFiles(globalContext, inputFiles);

  // Process input files into output
  const tempOutputDirectory = await fs.mkdtemp(nodePath.join(os.tmpdir(), 'phantomake-'));
  for (const inputFile of inputFiles) {
    // Dot directories / dotfiles are not processed
    if (inputFile.parsedRelativePath.name.startsWith('.') || inputFile.isWithinDotDirectory) {
      continue;
    }

    if (inputFile.isText) {
      const outputs = [await renderOutput(inputFile, globalContext, templates)];

      // If a paginator was created, we have to output multiple files
      const paginator = globalContext.paginators[inputFile.path];
      if (paginator) {
        for (const page of paginator.pages) {
          paginator.currentPage = page.number;
          const output = await renderOutput(inputFile, globalContext, templates);
          output.path = page.outputPath;
          outputs.push(output);
        }
      }

      if (outputs.length === 1) {
        logger?.verbose(`${inputFile.relativePath} ──> ${outputs[0].path}`);
      } else {
        logger?.verbose(inputFile.relativePath);
        for (const output of outputs) {
          logger?.verbose(` └─> ${output.path}`);
        }
      }

      // Write the final outputs
      for (const output of outputs) {
        const fullOutputPath = nodePath.join(tempOutputDirectory, output.path);
        await fs.mkdir(nodePath.dirname(fullOutputPath), { recursive: true });
        await Bun.write(fullOutputPath, output.content);
      }
    } else {
      // Non-text files are copied
      logger?.verbose(`${inputFile.relativePath} ──> ${inputFile.relativePath}`);
      const outputPath = nodePath.join(tempOutputDirectory, inputFile.relativePath);
      await fs.mkdir(nodePath.dirname(outputPath), { recursive: true });
      await Bun.write(outputPath, inputFile.file);
    }
  }

  // Write output
  await fs.cp(tempOutputDirectory, outputDirectory, { recursive: true });
}
