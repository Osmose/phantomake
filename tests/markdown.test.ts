import * as nodePath from 'node:path';
import * as fs from 'node:fs/promises';
import * as os from 'node:os';
import dedent from 'dedent';
import { expect, test, describe } from 'bun:test';
import phantomake from '../src';
import { walk } from '../src/util';

async function testInputFiles(label: string, contents: Record<string, string>) {
  test(label, async () => {
    const tmpDirectory = await fs.mkdtemp(nodePath.join(os.tmpdir(), 'phantomaketest-'));
    const inputDirectory = nodePath.join(tmpDirectory, 'input');
    const outputDirectory = nodePath.join(tmpDirectory, 'output');

    for (const [filename, content] of Object.entries(contents)) {
      await Bun.write(nodePath.join(inputDirectory, filename), content);
    }

    await phantomake(inputDirectory, outputDirectory);

    const outputPaths = await walk(outputDirectory);
    const outputContent: Record<string, string> = {};
    for (const outputPath of outputPaths) {
      outputContent[nodePath.relative(outputDirectory, outputPath)] = await Bun.file(outputPath).text();
    }

    expect(outputContent).toMatchSnapshot();
  });
}

describe('Markdown', () => {
  testInputFiles('Basic markdown', {
    'file.md': dedent`
      # Markdown Test file

      Testing if markdown output is correct.

      - Bulleted
      - List
    `,
    '.templates/default.ejs': `<%- output.content %>`,
  });

  testInputFiles('Include block', {
    'main.md': dedent`
      ::include{path=.markdown_include.md}

      The sum of 2 plus 2 is :include{path=.add.ejs arg1=2 arg2=2}
    `,
    '.markdown_include.md': dedent`
      # Heading

      Paragraph
    `,
    '.add.ejs': dedent`<%- Number.parseInt(arg1) + Number.parseInt(arg2) -%>`,
    '.templates/default.ejs': `<%- output.content %>`,
  });
});

describe('EJS', () => {
  testInputFiles('ctx.renderMarkdown', {
    'test.html.ejs': dedent`
      <%- await ctx.renderMarkdown("# Heading"); %>
    `,
  });
});
