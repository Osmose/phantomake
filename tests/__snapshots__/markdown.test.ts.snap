// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown Basic markdown 1`] = `
{
  "file.html": 
"<h1 id="markdown-test-file"><a href="#markdown-test-file">Markdown Test file</a></h1>
<p>Testing if markdown output is correct.</p>
<ul>
<li>Bulleted</li>
<li>List</li>
</ul>"
,
}
`;

exports[`Markdown Include block 1`] = `
{
  "main.html": 
"<h1 id="heading"><a href="#heading">Heading</a></h1>
<p>Paragraph</p>
<p>The sum of 2 plus 2 is 4</p>"
,
}
`;

exports[`EJS ctx.renderMarkdown 1`] = `
{
  "test.html": "<h1 id="heading"><a href="#heading">Heading</a></h1>",
}
`;
